spec/abci++/abci++_methods_002_draft.md

```diff
@@ -303,24 +304,26 @@ From the App's perspective, they'll probably skip ProcessProposal
     | Name                    | Type                                             | Description                                                                                 | Field Number |
     |-------------------------|--------------------------------------------------|---------------------------------------------------------------------------------------------|--------------|
     | modified_tx             | bool                                             | The Application sets it to true to denote it made changes to transactions                   | 1            |
-    | tx                      | repeated [TransactionRecord](#transactionrecord) | Possibly modified list of transactions that have been picked as part of the proposed block. | 2            |
+    | txs                     | repeated [TransactionRecord](#transactionrecord) | Possibly modified list of transactions that have been picked as part of the proposed block. | 2            |
     | app_hash                | bytes                                            | The Merkle root hash of the application state.                                              | 3            |
-    | tx_result               | repeated [TxResult](#txresult)                   | List of structures containing the data resulting from executing the transactions            | 4            |
+    | tx_results              | repeated [ExecTxResult](#txresult)               | List of structures containing the data resulting from executing the transactions            | 4            |
     | validator_updates       | repeated [ValidatorUpdate](#validatorupdate)     | Changes to validator set (set voting power to 0 to remove).                                 | 5            |
     | consensus_param_updates | [ConsensusParams](#consensusparams)              | Changes to consensus-critical gas, size, and other parameters.                              | 6            |
 
 * **Usage**:
     * Contains a preliminary block to be proposed, called _raw block_, which the Application can modify.
-    * The parameters and types of `RequestPrepareProposal` are the same as `RequestProcessProposal`
+    * The first five parameters of `RequestPrepareProposal` are the same as `RequestProcessProposal`
       and `RequestFinalizeBlock`.
     * The header contains the height, timestamp, and more - it exactly matches the
       Tendermint block header.
     * The Application can modify the transactions received in `RequestPrepareProposal` before sending
       them in `ResponsePrepareProposal`. In that case, `ResponsePrepareProposal.modified_tx` is set to true.
     * If `ResponsePrepareProposal.modified_tx` is false, then Tendermint will ignore the contents of
-      `ResponsePrepareProposal.tx`.
+      `ResponsePrepareProposal.txs`.
+    * If the Application modifies the transactions, the modified block MUST NOT exceed the configured maximum size,
```

Comment:
Can we give more specific guidance for what 'modified block' means here? Is it that `sizeof(txs)` in bytes cannot be greater than the `MaxBytes`? Or that all of the fields that the application returns cannot be greater than that? I think that passing the `max_bytes` along to the application makes sense, but I'm not yet sure it's obvious what the application should do with the `MaxBytes` if it is going to modify the block.

Reply:
Very good point William. The complication here is that, once we introduce vote extensions (v0.37), we may also give the App the chance to modify *app_signed* data.

After a bit of thinking, this is what I'm gonna do:

* I'll rename this field to *max_tx_bytes*
* We will add another field in v0.37: *max_extension_bytes*

Tendermint will calculate what those fields should be just before calling `RequestPrepareProposal`.

Let me know if you have concerns with this approach.

Reply:
I think that this seems very reasonable.